internal/stacks/stackruntime/internal/stackeval/component_config.go

```diff
@@ -408,6 +413,16 @@ func (c *ComponentConfig) ExprReferenceValue(ctx context.Context, phase EvalPhas
 	return cty.DynamicVal
 }
 
+func (c *ComponentConfig) ResolveExpressionReference(ctx context.Context, ref stackaddrs.Reference) (Referenceable, tfdiags.Diagnostics) {
+	repetition := instances.RepetitionData{}
+	if c.Declaration(ctx).ForEach != nil {
+		// For validation, we'll return unknown for the instance data.
```

Comment:
I don't know that it really matters that much since the logic is so trivial anyway, but FWIW there are some helper functions/variables in package `instances` for producing these "placeholder" repetition objects when we're validating before expansion. (The deferred actions logic makes use of them too, for example.)

Reply:
I might take a look in a follow up PR at simplifying this to use those shared functions. We're adding in refinements here that aren't present in the existing functions, and there's multiple places in this package that is doing the same. 

I could add an extra function in `instances` that also adds the unknown refinement, but I'll keep this PR as simple as possible without adding extra refactorings in that could be revisited later.